developer doc for ao2mo (#21)
pyscf/pyscf.github.io: source/develop/ao2mo_developer.rst

.. _developer_ao2mo:

***********************
AO to MO transformation
***********************

*Modules*: :mod:`ao2mo`

Overview
========
The :mod:`ao2mo` module exposes users with :func:`kernel`, :class:`load`, :func:`resture` to handle 4-index AO to MO transformation.
The function :func:`kernel` takes in :class:`Mole` object or the ao integrals stored in a :class:`numpy.ndarray` and a set of mo coefficients
If ``erifile`` is specified, the integrals will be stored in a HDF5 file, otherwise it will be stored in a numpy array.

> ::

Reviewer: Hi :)! I haven't read through this yet, but I see that this first example here at least is quite complete (with all the imports etc). Do you think it might make sense to move it into the example folder and import it?

Author: Probably, I was thinking of them as a single runnable script that shows all different usage, but it might be a duplicate of existing examples. Btw, does anyone know how are the examples indexed? The indices of existing examples seems rather random to me.

Reviewer: I think there are also very small examples, see e.g. https://github.com/pyscf/pyscf/blob/master/examples/cc/00-simple_ccsd.py . I am not sure how they are indexed to be honest. Probably starting from 00, 01, .... I have always assumed that sometimes examples are used in multiple directories so that is why the numbering scheme can make a jump but I don't actually know.

Reviewer 2: I'll merge this for the upcoming alpha release; can you add the examples in PySCF and `include` them in the docs in a future PR?

    from pyscf import gto, scf, ao2mo
    import tempfile
    mol = gto.Mole()
    mol.build(
        atom = 'H 0 0 0; F 0 0 1.1',  # in Angstrom
        basis = 'ccpvdz',
        symmetry = True,
    )

    mf = scf.RHF(mol)
    mf.kernel()

    orb = mf.mo_coeff

>     # eri is a numpy array

Reviewer 2: ```suggestion
    # get the two-electron integrals as a numpy array
```

    eri = ao2mo.kernel(mol, orb)

    ftmp = tempfile.NamedTemporaryFile()

>     # the integrals are stored in file ftmp

Reviewer 2: ```suggestion
    # saves the two-electron integrals in the file ftmp.name
```

.. _developer_ao2mo:

***********************
AO to MO transformation
***********************

*Modules*: :mod:`ao2mo`

Overview
========

> The :mod:`ao2mo` module exposes users to important functionality to handle transformations
> of 2-electron integrals (4-index AO to MO transformation) between different types of orbitals. The most important functions are :func:`kernel`, :class:`load`, :func:`restore`. 

Reviewer 2: ```suggestion
The :mod:`ao2mo` module exposes important functionality for transforming 2-electron integrals, i.e. the 4-index transformation from atomic orbitals (AOs) to molecular orbitals (MOs).
The most important functions are :func:`kernel`, :class:`load`, :func:`restore`. 
```

> The function :func:`kernel` takes in a :class:`Mole` object or the AO integrals stored in a :class:`numpy.ndarray` and a set of MO coefficients

Reviewer 2: ```suggestion
The function :func:`kernel` takes in a :class:`Mole` object or the AO integrals stored in a :class:`numpy.ndarray` and a set of MO coefficients.
```

> If ``erifile`` is specified, the integrals will be stored in a HDF5 file, otherwise it will be returned as a numpy array.

Reviewer 2: ```suggestion
If ``erifile`` is specified, the integrals will be stored in a HDF5 file with the given name, otherwise the integrals are returned as a numpy array.
```

The dataname can also be taken as a second argument.
In this way, it helps users access integrals stored in memory and in a file on the same footing.
The :class:`load` class should only be used within a ``with`` statement.
::
    eri = ao2mo.kernel(mol, orb)
    print(type(eri))
    with ao2mo.load(erifile) as eri:
        print(eri.shape)
    
    ao2mo.kernel(mol, orb, erifile = 'hf.h5', dataname = 'test')
    # load 2e integrals by filename and dataname
    with ao2mo.load('hf.h5', 'test') as eri:
        print(eri.shape)
    
    import numpy
    erirand = numpy.random.random((5,5,5,5))
    # load 2e integrals from numpy array
    with ao2mo.load(erirand) as eri:
        print(eri.shape)

Transform integrals between symmetries
======================================
The function :func:`restore` can transform 2e integrals or any 4-index quantity (e.g. 2rdm) between different permutation symmetries.
It takes a symmetry label, a ``numpy.ndarray`` as the 4-index quantity and the dimension ``norb`` of this quantity (the four indices
must have the same dimension).

> The symmetry label specifies the outpu symmetry the user desires, while

Reviewer 2: ```suggestion
The symmetry label specifies the output symmetry the user desires, while
```

the symmetry of the input array is determined by its shape.
The relation between different symmetries and shape is described below.

"Plain" MO integrals or 's1' symmetry have a shape of (norb, norb, norb, norb).
If a pair of indices have permutation symmetry, then only the triangular index is used and the shape is (npair, npair) with npair = norb*(norb+1)/2.
Thus integrals with 's2ij' and 's2kl' symmetries have a shape of (npair, norb, norb) or (norb, norb, npair) respectively.
If the integrals have 's4' symmetry, then there is permutation symmetry between both ij indices and kl indices and the shape will be (npair, npair).

> If the integral have 's8' symmetry, then permutation symmetry also exists between the ij pair and kl pair.

Reviewer 2: ```suggestion
If the integrals have 's8' symmetry, then permutation symmetry also exists between the ij pair and kl pair.
```
